src/ChannelServer/World/Character.cs

```diff
@@ -555,6 +639,392 @@ public int MaxMAtk
+
+				var byLevel = Level / 4f;
+				var byStat = (stat / 2f) + ((float)Math.Floor(stat / 15f) * 3f);
+				var byItem = 0; // TODO: Cached Accuracy for inventory/equip
+
+				var value = byLevel + byStat + byItem;
+
+				// Buffs: "HR_BM"
+				var byBuffs = 0;
+
+				// Rate buffs: Does tos have something like Accuracy + x% ?
+				var rate = 0;
+				var byRateBuffs = (float)Math.Floor(value * rate);
+
+				value += byBuffs + byRateBuffs;
+
+				return (int)value;
+			}
+		}
+
+		/// <summary>
+		/// Returns Evasion.
+		/// </summary>
+		public int Evasion
+		{
+			get
+			{
+				var Level = this.Level;
+				var stat = this.Dex;
+
+				var byLevel = Level / 4f;
+				var byStat = (stat / 2f) + ((float)Math.Floor(stat / 15f) * 3f);
+				var byItem = 0; // TODO: Cached Evasion for inventory/equip
+
+				var value = byLevel + byStat + byItem;
+
+				//if(hasBuff("Finestra"))
+				//	if(hasAttribute("Finestra: Splash")) | reduces evasion rate by x2 more of the existing value when [Finestra] is active
+				//		value -= (15 + ((SkillLevel - 1) * 15)) * 2
+				//	else
+				//		value -= 15 + ((SkillLevel - 1) * 15)
+
+
+
+				// Buffs: "DR_BM"
+				var byBuffs = 0;
+
+				// Rate buffs: Can't find the right property
+				//if(hasBuff("Evasive Action"))
+				//	rate += SkillLevel*2
+				var rate = 0;
+				var byRateBuffs = (float)Math.Floor(value * rate);
+
+				value += byBuffs + byRateBuffs;
+
+				return (int)value;
+			}
+		}
+
+		/// <summary>
+		/// Returns Block.
+		/// </summary>
+		public int Block
+		{
+			get
+			{
+				var byItem = 0f; // TODO: Cached Block for inventory/equip ("BLK")
+
+				var value = byItem;
+
+				//if(hasBuff("High Guard"))
+				//	value += 100 + ((SkillLevel - 1) * 30)
+				//if(hasBuff("Finestra"))
+				//	value += 40 + ((SkillLevel - 1) * 15)
+
+				// Buffs: "BLK_BM"
+				var byBuffs = 0;
+
+				// Rate buffs:
+				var rate = 0;
+				var byRateBuffs = (float)Math.Floor(value * rate);
+
+				value += byBuffs + byRateBuffs;
+
+				return (int)value;
+			}
+		}
+
+		/// <summary>
+		/// Returns Block Penetration.
+		/// </summary>
+		public int BlockPen
+		{
+			get
+			{
+				var Level = this.Level;
+				var stat = this.Dex;
+
+				var byLevel = Level / 4f;
+				var byStat = (stat / 2f) + ((float)Math.Floor(stat / 15f) * 3f);
+				var byItem = 0; // TODO: Cached Block Penetration for inventory/equip ("BLK_BREAK")
+
+				var value = byLevel + byStat + byItem;
+
+				// Buffs: "BLK_BREAK_BM"
+				var byBuffs = 0;
+
+				// Rate buffs: 
+				var rate = 0;
+				var byRateBuffs = (float)Math.Floor(value * rate);
+
+				value += byBuffs + byRateBuffs;
+
+				return (int)value;
+			}
+		}
+
+		/// <summary>
+		/// Returns Aoe Attack Ratio.
+		/// </summary>
+		public int AoeRatio
+		{
+			get
+			{
+				var baseValue = 4;
+				var byItem = 0f; // TODO: Cached Aoe Attack Ratio for inventory/equip  ("SR")
+
+				var value = baseValue + byItem;
+
+				//if(hasBuff("Finestra") && hasAttribute("Finestra: Splash"))
```

Comment:
This is most-likely not how the buffs are gonna work. For example, the buff effects for "Finestra" would be added to the character's list of buffs on activation, which are then added to the value via byBuffs and other variables. I only included those hasBuff comments in the previous properties for future reference, because it was hardcoded in the client files as well.

It would be pretty annoying if we had to hardcode every single buff like this^^